File: Java/Hard/WrongAnswer/number-of-ways-to-reorder-array-to-get-same-bst_1.java
class Solution {
    public int numOfWays(int[] nums) {
        
        return (int) (helper(nums, 0, nums.length - 1) - 1);
    }

    private long helper(int[] nums, int start, int end) {
        if (start >= end) return 1;

        int root = nums[start];
        int left = start + 1;
        int right = end;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < root) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        int leftSize = left - start - 1;
        int rightSize = end - left + 1;

        return (helper(nums, start + 1, left - 1) * helper(nums, left, end) * (leftSize + rightSize + 1)) % 1_000_000_007;
    }
}

File: C++/Hard/RuntimeError/number-of-ways-to-reorder-array-to-get-same-bst_1.cpp
class Solution {
public:
    int numOfWays(vector<int>& nums) {
        
        int n = nums.size();
        if (n <= 2) return 1;
        int root = nums[0];
        vector<int> left, right;
        for (int i = 1; i < n; ++i) {
            if (nums[i] < root) left.push_back(nums[i]);
            else right.push_back(nums[i]);
        }
        long long res = (long long)numOfWays(left) * numOfWays(right) % 1000000007;
        for (int i = 1; i < n; ++i) {
            if (nums[i] > root) break;
            left.push_back(nums[i]);
            right.erase(find(right.begin(), right.end(), nums[i]));
            res = (res + (long long)numOfWays(left) * numOfWays(right)) % 1000000007;
        }
        return res;
    }
};